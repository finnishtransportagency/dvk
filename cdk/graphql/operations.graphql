query findFairways {
  fairways {
    ...fragment FairwayFragment on Fairway {
      id
      nameFI
      nameSV
    }
  }
}

query findFairway($id: Int) {
  fairway(id: $id) {
    ...Fairway
  }
}
