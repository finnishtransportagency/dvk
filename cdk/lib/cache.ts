export type CloudFrontCacheKey =
  | 'aislocations'
  | 'aisvessels'
  | 'mareograph'
  | 'observation'
  | 'buoy'
  | 'marinewarning'
  | 'feature'
  | 'pilotroutes'
  | 'forecast';

export const FEATURE_CACHE_DURATION = 7200; // 2 hours

type CloudFrontCacheItem = {
  key: CloudFrontCacheKey;
  maxAge: number;
  staleWhileRevalidate: number;
  staleIfError: number;
};

const CloudFrontCacheItems: Array<CloudFrontCacheItem> = [
  {
    key: 'aislocations',
    maxAge: 2, // 2 seconds
    staleWhileRevalidate: 3, // 3 seconds
    staleIfError: 58, // 58 seconds (do not return location data more than 1 minute old)
  },
  {
    key: 'aisvessels',
    maxAge: 300, // 5 minutes
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 300, // 5 minutes (do not return vessel data more than 10 minutes old)
  },
  {
    key: 'mareograph',
    maxAge: 60, // 1 minute
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
  {
    key: 'observation',
    maxAge: 60, // 1 minute
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
  {
    key: 'buoy',
    maxAge: 60, // 1 minute
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
  {
    key: 'marinewarning',
    maxAge: 540, // 9 minutes
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
  {
    key: 'feature',
    maxAge: 7140, // 1 hour 59 minutes
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
  {
    key: 'pilotroutes',
    maxAge: 7140, // 1 hour 59 minutes
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
  {
    key: 'forecast',
    maxAge: 7140, // 1 hour 59 minutes
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
];

export function getAisCacheControlHeaders(key: 'aisvessels' | 'aislocations'): Record<string, string[]> {
  return getCloudFrontCacheControlHeaders(key);
}

export function isCloudFrontCacheKey(key: string): key is CloudFrontCacheKey {
  return CloudFrontCacheItems.some((item) => item.key === key );
}

export function getFeatureCacheControlHeaders(): Record<string, string[]> {
  return getCloudFrontCacheControlHeaders('feature');
}

export function getCloudFrontCacheControlHeaders(key: CloudFrontCacheKey): Record<string, string[]> {
  const cacheItem = CloudFrontCacheItems.find((item) => item.key === key );
  if (!cacheItem) {
    return getFeatureCacheControlHeaders();
  }
  const maxAge = cacheItem.maxAge;
  const staleWhileRevalidate = cacheItem.staleWhileRevalidate;
  const staleIfError = cacheItem.staleIfError;

  return {
    'Cache-Control': ['max-age=' + maxAge + ', ' + 'stale-while-revalidate=' + staleWhileRevalidate + ', ' + 'stale-if-error=' + staleIfError],
  };
}
