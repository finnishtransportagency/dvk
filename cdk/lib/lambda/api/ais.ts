import { Feature, FeatureCollection, GeoJsonProperties, Geometry } from 'geojson';
import { Vessel, VesselAPIModel, VesselLocationFeatureCollection } from './apiModels';
import { fetchAISFeatureCollection, fetchAISMetadata } from './axios';

const aisV1Path = 'ais/api/ais/v1/';

function getTimestampParams(): Record<string, string> {
  const now = Date.now();
  const yesterday = Date.now() - 24 * 60 * 60 * 1000;
  return { from: yesterday.toString(), to: now.toString() };
}

function parseFrom64BitInteger(integer: number, bitStart: number, bitEnd: number) {
  // 64-bit integer: get bit range from start to end by losing other bits
  // Shift bits first to left and then to right, let others fall off
  const i = (integer << (63 - bitStart)) >>> (63 - bitStart + bitEnd);
  return i.toString().padStart(2, '0');
}

/* MMDDHHMM UTC
Bits 19-16: month; 1-12; 0 = not available = default
Bits 15-11: day; 1-31; 0 = not available = default
Bits 10-6: hour; 0-23; 24 = not available = default
Bits 5-0: minute; 0-59; 60 = not available = default
*/
function parseETA(eta: number): string {
  const month = parseFrom64BitInteger(eta, 19, 16);
  const day = parseFrom64BitInteger(eta, 15, 11);
  const hour = parseFrom64BitInteger(eta, 10, 6);
  const minute = parseFrom64BitInteger(eta, 5, 0);
  return month + '-' + day + ' ' + hour + ':' + minute;
}

export async function fetchVessels(): Promise<Vessel[]> {
  const path = aisV1Path + 'vessels';
  const params = getTimestampParams();
  const data: VesselAPIModel[] = await fetchAISMetadata(path, params);
  return data.map((row) => {
    return {
      name: row.name,
      timestamp: new Date(row.timestamp),
      mmsi: row.mmsi,
      callSign: row.callSign,
      imo: row.imo,
      shipType: row.shipType,
      draught: row.draught / 10,
      eta: parseETA(row.eta),
      posType: row.posType,
      referencePointA: row.referencePointA,
      referencePointB: row.referencePointB,
      referencePointC: row.referencePointC,
      referencePointD: row.referencePointD,
      destination: row.destination,
    };
  });
}

export async function fetchLocations(): Promise<FeatureCollection> {
  const path = aisV1Path + 'locations';
  const params = getTimestampParams();
  const data: VesselLocationFeatureCollection = await fetchAISFeatureCollection(path, params);
  const { dataUpdatedTime, features } = data;
  const geojsonFeatures: Feature<Geometry, GeoJsonProperties>[] = features.map((f) => {
    return {
      type: 'Feature',
      id: f.mmsi,
      geometry: f.geometry,
      properties: {
        ...f.properties,
        dataUpdatedTime,
        timestampExternal: new Date(f.properties.timestampExternal),
        featureType: 'aisvessel',
      },
    };
  });
  const collection: FeatureCollection = {
    type: 'FeatureCollection',
    features: geojsonFeatures,
  };
  return collection;
}
