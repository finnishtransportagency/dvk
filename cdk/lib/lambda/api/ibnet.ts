import { Feature, FeatureCollection, GeoJsonProperties, Geometry } from 'geojson';
import { Dirway, DirwayPoint, IBNetApiResponse } from './apiModels';
import { fetchIBNetApi } from './axios';
import { log } from '../logger';
import { Position } from '@turf/helpers';

export const DIRWAY_PATH = 'dirway';
export const DIRWAY_POINT_PATH = 'dirwaypoint';

function addDirwayFeatures(features: Feature<Geometry, GeoJsonProperties>[], dirwayPoints: DirwayPoint[], dirways: Dirway[]) {
  const dirwayMap = new Map<string, DirwayPoint[]>();
  for (const p of dirwayPoints) {
    if (!p.deleted) {
      dirwayMap.set(p.dirway_id, (dirwayMap.get(p.dirway_id) ?? []).concat(p));
    }
  }
  for (const points of dirwayMap.values()) {
    const id = points[0].dirway_id;
    const coordinates = points.sort((a, b) => a.order_num - b.order_num).map((p) => [p.longitude, p.latitude] as Position);
    const dirway = dirways.filter((d) => !d.deleted).find((d) => d.id === id);

    const properties = {
      id: id,
      name: dirway?.name,
      description: dirway?.description,
      updated: dirway?.change_time,
      featureType: 'dirway',
    };

    features.push({
      type: 'Feature',
      id: id,
      geometry: { type: 'LineString', coordinates: coordinates },
      properties: properties,
    });
  }
}

export async function fetchDirways(): Promise<FeatureCollection> {
  const features: Feature<Geometry, GeoJsonProperties>[] = [];

  // First fetch root of the API to get max toRV value
  const apiResponse: IBNetApiResponse | null = await fetchIBNetApi<IBNetApiResponse>();
  const { toRv } = apiResponse ?? {};

  if (toRv) {
    const params = { from: '0', to: toRv };
    const dirways: Dirway[] | null = await fetchIBNetApi<Dirway[]>(DIRWAY_PATH, params);
    const dirwayPoints: DirwayPoint[] | null = await fetchIBNetApi<DirwayPoint[]>(DIRWAY_POINT_PATH, params);

    if (dirways && dirwayPoints) {
      addDirwayFeatures(features, dirwayPoints, dirways);
    }
  } else {
    log.error('Could not get toRv from IBNetApi, not fetching dirway data');
  }

  const collection: FeatureCollection = {
    type: 'FeatureCollection',
    features: features,
  };
  return collection;
}
