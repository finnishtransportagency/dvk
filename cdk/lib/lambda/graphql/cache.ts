import { DeleteObjectsCommand, GetObjectCommand, ObjectIdentifier, PutObjectCommand, S3Client } from '@aws-sdk/client-s3';
import { getEnvironment } from '../environment';
import { log } from '../logger';
import { Readable } from 'stream';

const s3Client = new S3Client({ region: 'eu-west-1' });

const AIS_LOCATION_CACHE = {
  MAX_AGE: 2, // 2 seconds
  STALE_WHILE_REVALIDATE: 3, // 3 seconds
  STALE_IF_ERROR: 58, // 58 seconds (do not return location data more than 1 minute old)
};
const AIS_VESSEL_CACHE = {
  MAX_AGE: 300, // 5 minutes
  STALE_WHILE_REVALIDATE: 60, // 1 minute
  STALE_IF_ERROR: 300, // 5 minutes (do not return vessel data more than 10 minutes old)
};

export const FEATURE_CACHE_DURATION = 7200; // 2 hours

function getCacheBucketName() {
  return `featurecache-${getEnvironment()}`;
}

export function getFeatureCacheDuration(key: string) {
  if (key === 'aislocations') {
    return AIS_LOCATION_CACHE.MAX_AGE;
  } else if (key === 'aisvessels') {
    return AIS_VESSEL_CACHE.MAX_AGE;
  } else {
    return FEATURE_CACHE_DURATION;
  }
}

export function getCacheControlHeaders(key: string): Record<string, string[]> {
  if (key === 'aislocations') {
    return {
      'Cache-Control': [
        'max-age=' +
          AIS_LOCATION_CACHE.MAX_AGE +
          ', ' +
          'stale-while-revalidate=' +
          AIS_LOCATION_CACHE.STALE_WHILE_REVALIDATE +
          ', ' +
          'stale-if-error=' +
          AIS_LOCATION_CACHE.STALE_IF_ERROR,
      ],
    };
  } else if (key === 'aisvessels') {
    return {
      'Cache-Control': [
        'max-age=' +
          AIS_VESSEL_CACHE.MAX_AGE +
          ', ' +
          'stale-while-revalidate=' +
          AIS_VESSEL_CACHE.STALE_WHILE_REVALIDATE +
          ', ' +
          'stale-if-error=' +
          AIS_VESSEL_CACHE.STALE_IF_ERROR,
      ],
    };
  } else {
    return {};
  }
}

export async function cacheResponse(key: string, response: object | string) {
  const cacheDurationSeconds = getFeatureCacheDuration(key);
  const expires = new Date();
  expires.setTime(expires.getTime() + cacheDurationSeconds * 1000);
  const command = new PutObjectCommand({
    Key: key,
    Bucket: getCacheBucketName(),
    Expires: expires,
    Body: typeof response === 'string' ? response : JSON.stringify(response),
  });
  await s3Client.send(command);
}

export type CacheResponse = {
  expired: boolean;
  data?: string;
  lastModified?: string;
};

export async function getFromCache(key: string): Promise<CacheResponse> {
  try {
    const data = await s3Client.send(
      new GetObjectCommand({
        Key: key,
        Bucket: getCacheBucketName(),
      })
    );
    if (data.Body) {
      log.debug(`returning ${key} from cache`);
      const streamToString = (stream: Readable): Promise<string> =>
        new Promise((resolve, reject) => {
          const chunks: Uint8Array[] = [];
          stream.on('data', (chunk: Uint8Array) => chunks.push(chunk));
          stream.on('error', reject);
          stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
        });
      const expires = data.ExpiresString ? Date.parse(data.ExpiresString) : undefined;
      return {
        expired: expires !== undefined && !isNaN(expires) && expires < Date.now(),
        data: await streamToString(data.Body as Readable),
        lastModified: data.LastModified?.toUTCString(),
      };
    }
  } catch (e) {
    // errors ignored also not found
  }
  return { expired: true };
}

export async function deleteCacheObjects(keys: string[]) {
  const objects: ObjectIdentifier[] = keys.map((key) => {
    return { Key: key };
  });
  const response = await s3Client.send(
    new DeleteObjectsCommand({
      Bucket: getCacheBucketName(),
      Delete: {
        Objects: objects,
      },
    })
  );
  log.debug(`Deleted cache objects ${keys}: ${response.Deleted?.length}/${keys.length}`);
  if (response.Errors?.length) {
    log.error(response.Errors);
  }
}
