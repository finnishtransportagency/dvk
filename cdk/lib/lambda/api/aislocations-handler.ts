import { ALBEvent, ALBResult } from 'aws-lambda';
import { log } from '../logger';
import { fetchLocations } from './ais';
import { handleAisCall } from '../util';

export const handler = async (event: ALBEvent): Promise<ALBResult> => {
  log.info({ event }, `aislocations()`);
  return handleAisCall('aislocations', fetchLocations, ['application/geo+json']);
};
