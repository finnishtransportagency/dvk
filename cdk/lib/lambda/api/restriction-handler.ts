import { ALBEvent, ALBResult } from 'aws-lambda';
import { log } from '../logger';
import { toBase64Response } from '../util';
import { getHeaders } from '../environment';
import { fetchRestrictions } from './ibnet';
import { getFeatureCacheControlHeaders } from '../../cache';

export const RESTRICTIONS_KEY = 'restrictions';

export const handler = async (event: ALBEvent): Promise<ALBResult> => {
  log.info({ event }, `restrictions()`);
  let base64Response: string | undefined;
  let statusCode = 200;
  try {
    const restrictionData = await fetchRestrictions();
    log.debug('restriction data: %d', restrictionData.features.length);
    base64Response = await toBase64Response(restrictionData);
  } catch (e) {
    base64Response = undefined;
    statusCode = 503;
  }

  return {
    statusCode,
    body: base64Response,
    isBase64Encoded: true,
    multiValueHeaders: {
      ...getHeaders(),
      ...getFeatureCacheControlHeaders(),
      'Content-Type': ['application/geo+json'],
    },
  };
};
