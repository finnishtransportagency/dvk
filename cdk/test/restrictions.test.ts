import { mockALBEvent } from './mocks';
import { gunzip } from 'zlib';
import assert from 'assert';
import { FeatureCollection } from 'geojson';
import { getFeatureCacheControlHeaders } from '../lib/lambda/graphql/cache';
import { RESTRICTIONS_KEY, handler } from '../lib/lambda/api/restriction-handler';
import { LOCATION_PATH, RESTRICTION_PATH, Restriction } from '../lib/lambda/api/ibnet';

jest.mock('../lib/lambda/environment', () => ({
  getEnvironment: () => 'mock',
  isPermanentEnvironment: () => false,
  getHeaders: () => {},
}));

const apiRootResponse = {
  rvEndpoints: ['icebreaker_status', 'dirwaypoint', 'location', 'dirway', 'restriction', 'source', 'aistarget', 'queue', 'activity', 'vessel'],
  toRv: 9420308526,
};

const locationResponse = [
  {
    rv: 7894335543,
    change_time: '2021-12-16T11:03:24.460Z',
    id: 'location-2_1041',
    type: 'PORT',
    name: 'RUNDVIK',
    locode_list: 'SERUV',
    nationality: 'SE',
    latitude: 63.5333,
    longitude: 19.45,
    winterport: true,
  },
  {
    rv: 7895851919,
    change_time: '2022-11-18T00:00:00Z',
    id: 'location-2_931',
    type: 'PORT',
    name: 'HAMINA',
    locode_list: 'FIHMN',
    nationality: 'FI',
    latitude: 60.5667,
    longitude: 27.1833,
    winterport: true,
  },
  {
    rv: 7895851919,
    change_time: '2023-12-19T00:00:00Z',
    id: 'location-2_931',
    type: 'PORT',
    name: 'HAMINA',
    locode_list: 'FIHMN',
    nationality: 'FI',
    latitude: 60.5667,
    longitude: 27.1833,
    winterport: true,
  },
  {
    rv: 7894335589,
    change_time: '2021-12-16T11:03:24.460Z',
    id: 'location-2_4081177',
    type: 'PORT',
    name: 'ORRSKÃ„R',
    locode_list: 'SEALA',
    nationality: 'SE',
    latitude: 61.22,
    longitude: 17.1717,
    winterport: true,
  },
  {
    rv: 7895795043,
    change_time: '2023-07-14T11:03:46.529Z',
    deleted: true,
    id: 'location-94_10004',
  },
  {
    rv: 7895851920,
    change_time: '2024-01-10T08:09:07.777Z',
    id: 'location-2_933',
    type: 'PORT',
    name: 'HANKO',
    locode_list: 'FIHKO',
    nationality: 'FI',
    latitude: 59.8167,
    longitude: 22.9667,
    winterport: true,
  },
  {
    rv: 7894335500,
    change_time: '2021-12-16T11:03:24.460Z',
    id: 'location-2_955',
    type: 'PORT',
    name: 'KANTVIK',
    locode_list: 'FIKNT',
    nationality: 'FI',
    latitude: 60.0833,
    longitude: 24.4,
    winterport: true,
  },
  {
    rv: 7894334131,
    change_time: '2021-12-16T11:03:24.460Z',
    deleted: true,
    id: 'location-94_100',
  },
  {
    rv: 7895851922,
    change_time: '2022-11-18T00:00:00Z',
    id: 'location-2_967',
    type: 'PORT',
    name: 'KASKINEN',
    locode_list: 'FIKAS',
    nationality: 'FI',
    latitude: 62.3833,
    longitude: 21.2167,
    winterport: true,
  },
  {
    rv: 7895851922,
    change_time: '2022-10-17T00:00:00Z',
    id: 'location-2_967',
    type: 'PORT',
    name: 'KASKINEN',
    locode_list: 'FIKAS',
    nationality: 'FI',
    latitude: 62.3833,
    longitude: 21.2167,
    winterport: true,
  },
  {
    rv: 7895851923,
    change_time: '2022-11-18T00:00:00Z',
    id: 'location-2_969',
    type: 'PORT',
    name: 'KEMI',
    locode_list: 'FIKEM',
    nationality: 'FI',
    latitude: 65.7333,
    longitude: 24.5667,
    winterport: true,
  },
  {
    rv: 7894334240,
    change_time: '2021-12-16T11:03:24.700Z',
    id: 'location-94_10001',
    type: 'FAIRWAY',
    name: 'OULU DW-ROUTE',
    nationality: 'FI',
    latitude: 65.2227,
    longitude: 24.674,
    winterport: true,
  },
];

const restrictionResponse = [
  {
    rv: 7895711037,
    change_time: '2023-02-27T09:16:27.067Z',
    id: 'restriction-2_35000041',
    location_id: 'location-2_1041',
    start_time: '2022-12-20T23:00:00Z',
    end_time: '2023-03-03T23:00:00Z',
    text_compilation: 'II 2000',
  },
  {
    rv: 7895711064,
    change_time: '2023-04-24T07:56:24.837Z',
    id: 'restriction-2_35000136',
    location_id: 'location-2_1041',
    start_time: '2023-03-03T23:00:00Z',
    end_time: '2023-04-23T22:00:00Z',
    text_compilation: 'I 2000',
  },
  {
    rv: 7895711080,
    change_time: '2023-05-02T07:02:55.195Z',
    id: 'restriction-2_35000178',
    location_id: 'location-2_1041',
    start_time: '2023-04-23T22:00:00Z',
    text_compilation: 'II 2000',
  },
  {
    rv: 7895711119,
    change_time: '2023-03-03T07:46:27.865Z',
    id: 'restriction-47_35000059',
    location_id: 'location-2_931',
    start_time: '2022-12-23T22:00:00Z',
    end_time: '2023-03-07T21:59:59Z',
    text_compilation: 'II 2000',
  },
  {
    rv: 7895711180,
    change_time: '2023-04-13T07:13:35.706Z',
    id: 'restriction-47_35000152',
    location_id: 'location-2_931',
    start_time: '2023-03-07T22:00:00Z',
    end_time: '2023-04-11T20:59:59Z',
    text_compilation: 'I 2000',
  },
  {
    rv: 7895711197,
    change_time: '2023-04-16T08:46:23.892Z',
    id: 'restriction-47_35000173',
    location_id: 'location-2_931',
    start_time: '2023-04-11T21:00:00Z',
    text_compilation: 'II 2000 -1',
  },
  {
    rv: 7895711066,
    change_time: '2023-04-03T07:32:47.463Z',
    id: 'restriction-2_35000138',
    location_id: 'location-2_4081177',
    start_time: '2023-03-05T23:00:00Z',
    end_time: '2023-04-02T22:00:00Z',
    text_compilation: 'II 2000',
  },
  {
    rv: 7895711161,
    change_time: '2023-04-05T07:46:32.792Z',
    id: 'restriction-47_35000105',
    location_id: 'location-2_955',
    start_time: '2023-01-06T22:00:00Z',
    end_time: '2023-04-04T21:00:00Z',
    text_compilation: 'II 2000',
  },
  {
    rv: 7895711157,
    change_time: '2023-04-12T07:29:36.860Z',
    id: 'restriction-47_35000101',
    location_id: 'location-2_967',
    start_time: '2023-01-06T22:00:00Z',
    text_compilation: 'II 2000',
  },
  {
    rv: 7895711092,
    change_time: '2022-12-19T08:25:01.995Z',
    id: 'restriction-47_35000001',
    location_id: 'location-2_969',
    start_time: '2022-11-30T22:00:00Z',
    text_compilation: 'II 2000',
  },
  {
    rv: 7895711151,
    change_time: '2023-01-27T07:24:33.406Z',
    id: 'restriction-47_35000095',
    location_id: 'location-2_969',
    start_time: '2023-01-06T22:00:00Z',
    text_compilation: 'IB 2000',
  },
  {
    rv: 7895711117,
    change_time: '2023-01-02T08:56:58.260Z',
    id: 'restriction-47_35000057',
    location_id: 'location-2_969',
    start_time: '2022-12-23T22:00:00Z',
    text_compilation: 'I 2000',
  },
  {
    rv: 7895711166,
    change_time: '2023-05-05T06:59:02.348Z',
    id: 'restriction-47_35000122',
    location_id: 'location-2_969',
    start_time: '2023-02-21T22:00:00Z',
    text_compilation: 'IA 4000',
  },
  {
    rv: 7895711163,
    change_time: '2023-02-17T08:23:36.416Z',
    id: 'restriction-47_35000114',
    location_id: 'location-2_969',
    start_time: '2023-01-31T22:00:00Z',
    text_compilation: 'IA 2000',
  },
  {
    rv: 7963559591,
    change_time: '2024-01-17T13:27:34.087Z',
    deleted: true,
    id: 'restriction-47_36000001',
  },
  {
    rv: 8744148057,
    change_time: '2024-04-08T07:58:38.719Z',
    id: 'restriction-47_36000003',
    location_id: 'location-2_969',
    start_time: '2024-01-13T22:00:00Z',
    text_compilation: 'IAS 4000',
  },
  {
    rv: 9345937317,
    change_time: '2024-05-30T08:29:26.031Z',
    id: 'restriction-47_36000037',
    location_id: 'location-94_10001',
    start_time: '2024-05-28T21:00:00Z',
    text_compilation: 'IA 3000',
  },
  {
    rv: 7895711180,
    change_time: '2023-04-12T07:13:35.706Z',
    id: 'restriction-47_35000152',
    location_id: 'location-2_931',
    start_time: '2023-03-07T22:00:00Z',
    end_time: '2023-04-11T20:59:59Z',
    text_compilation: 'I 2000 -1',
  },
  {
    rv: 7895711197,
    change_time: '2023-04-17T08:46:23.892Z',
    id: 'restriction-47_35000173',
    location_id: 'location-2_931',
    start_time: '2023-04-11T21:00:00Z',
    text_compilation: 'II 2000',
  },
  {
    rv: 7895711157,
    change_time: '2023-04-12T07:40:36.860Z',
    id: 'restriction-47_35000101',
    location_id: 'location-2_967',
    start_time: '2023-01-06T22:00:00Z',
    text_compilation: 'II 2000',
  },
];

async function parseResponse(body: string): Promise<FeatureCollection> {
  const response = new Promise<Error | Buffer>((resolve, reject) =>
    gunzip(Buffer.from(body, 'base64'), (err, data) => {
      if (err) {
        reject(err);
      }
      resolve(data);
    })
  );
  return JSON.parse((await response).toString()) as FeatureCollection;
}

let throwError = false;
jest.mock('../lib/lambda/api/axios', () => ({
  fetchIBNetApi: (path?: string) => {
    if (throwError) {
      throw new Error('Fetching from IBNet api failed');
    }
    if (!path) {
      return apiRootResponse;
    } else if (path === LOCATION_PATH) {
      return locationResponse;
    } else if (path === RESTRICTION_PATH) {
      return restrictionResponse;
    }
    return null;
  },
}));

beforeEach(() => {
  jest.resetAllMocks();
  throwError = false;
});

it('should get restrictions from api', async () => {
  const response = await handler(mockALBEvent(RESTRICTIONS_KEY));
  assert(response.body);
  const responseObj = await parseResponse(response.body);

  // 12 locations minus 8: 2 duplicates, 2 other nationalities, 2 deleted, 1 fairway, 1 without restrictions
  expect(responseObj.features.length).toBe(4);

  // Check correct duplicates are filtered
  const haminaDuplicate = responseObj.features.find((f) => f.id === 'location-2_931');
  expect(haminaDuplicate?.properties?.updated).toBe('2023-12-19T00:00:00.000Z');
  expect(haminaDuplicate?.properties?.restrictions?.length).toBe(3);
  const haminaRestrictionDuplicate = haminaDuplicate?.properties?.restrictions?.find((r: Restriction) => r.id === 'restriction-47_35000173');
  expect(haminaRestrictionDuplicate?.updated).toBe('2023-04-17T08:46:23.892Z');

  const kaskinenDuplicate = responseObj.features.find((f) => f.id === 'location-2_967');
  expect(kaskinenDuplicate?.properties?.updated).toBe('2022-11-18T00:00:00.000Z');
  expect(kaskinenDuplicate?.properties?.restrictions?.length).toBe(1);
  expect(kaskinenDuplicate?.properties?.restrictions?.[0].updated).toBe('2023-04-12T07:40:36.860Z');

  expect(responseObj).toMatchSnapshot();
});

it('should get internal server error when api call fails', async () => {
  throwError = true;
  const response = await handler(mockALBEvent(RESTRICTIONS_KEY));
  expect(response.statusCode).toBe(503);
});

it('should return right cache headers', async () => {
  const response = await handler(mockALBEvent(RESTRICTIONS_KEY));
  assert(response.body);
  const headers = getFeatureCacheControlHeaders(RESTRICTIONS_KEY)?.['Cache-Control'];
  expect(response?.multiValueHeaders?.['Cache-Control']).toStrictEqual(headers);
});
