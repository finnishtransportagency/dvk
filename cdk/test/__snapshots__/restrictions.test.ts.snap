// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get restrictions from api 1`] = `
{
  "features": [
    {
      "geometry": {
        "coordinates": [
          24.5667,
          65.7333,
        ],
        "type": "Point",
      },
      "id": "location-2_969",
      "properties": {
        "featureType": "restrictionport",
        "id": "location-2_969",
        "name": "KEMI",
        "restrictions": [
          {
            "description": "IAS 4000",
            "startTime": "2024-01-13T22:00:00.000Z",
            "updated": "2024-04-08T07:58:38.719Z",
          },
          {
            "description": "IA 4000",
            "startTime": "2023-02-21T22:00:00.000Z",
            "updated": "2023-05-05T06:59:02.348Z",
          },
          {
            "description": "IA 2000",
            "startTime": "2023-01-31T22:00:00.000Z",
            "updated": "2023-02-17T08:23:36.416Z",
          },
          {
            "description": "IB 2000",
            "startTime": "2023-01-06T22:00:00.000Z",
            "updated": "2023-01-27T07:24:33.406Z",
          },
          {
            "description": "I 2000",
            "startTime": "2022-12-23T22:00:00.000Z",
            "updated": "2023-01-02T08:56:58.260Z",
          },
          {
            "description": "II 2000",
            "startTime": "2022-11-30T22:00:00.000Z",
            "updated": "2022-12-19T08:25:01.995Z",
          },
        ],
        "updated": "2022-11-18T00:00:00.000Z",
      },
      "type": "Feature",
    },
    {
      "geometry": {
        "coordinates": [
          27.1833,
          60.5667,
        ],
        "type": "Point",
      },
      "id": "location-2_931",
      "properties": {
        "featureType": "restrictionport",
        "id": "location-2_931",
        "name": "HAMINA",
        "restrictions": [
          {
            "description": "II 2000",
            "startTime": "2023-04-11T21:00:00.000Z",
            "updated": "2023-04-17T08:46:23.892Z",
          },
          {
            "description": "I 2000",
            "endTime": "2023-04-11T20:59:59.000Z",
            "startTime": "2023-03-07T22:00:00.000Z",
            "updated": "2023-04-13T07:13:35.706Z",
          },
          {
            "description": "II 2000",
            "endTime": "2023-03-07T21:59:59.000Z",
            "startTime": "2022-12-23T22:00:00.000Z",
            "updated": "2023-03-03T07:46:27.865Z",
          },
        ],
        "updated": "2023-12-19T00:00:00.000Z",
      },
      "type": "Feature",
    },
    {
      "geometry": {
        "coordinates": [
          21.2167,
          62.3833,
        ],
        "type": "Point",
      },
      "id": "location-2_967",
      "properties": {
        "featureType": "restrictionport",
        "id": "location-2_967",
        "name": "KASKINEN",
        "restrictions": [
          {
            "description": "II 2000",
            "startTime": "2023-01-06T22:00:00.000Z",
            "updated": "2023-04-12T07:40:36.860Z",
          },
        ],
        "updated": "2022-11-18T00:00:00.000Z",
      },
      "type": "Feature",
    },
    {
      "geometry": {
        "coordinates": [
          24.4,
          60.0833,
        ],
        "type": "Point",
      },
      "id": "location-2_955",
      "properties": {
        "featureType": "restrictionport",
        "id": "location-2_955",
        "name": "KANTVIK",
        "restrictions": [
          {
            "description": "II 2000",
            "endTime": "2023-04-04T21:00:00.000Z",
            "startTime": "2023-01-06T22:00:00.000Z",
            "updated": "2023-04-05T07:46:32.792Z",
          },
        ],
        "updated": "2021-12-16T11:03:24.460Z",
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;
